File: DSA/NiharikaChand_590017697/Day19Question1.java
import java.util.*;
class Day19Question1 
{
    public static int search(int arr[], int n, int t) 
    {
        int l = 0, h = n - 1;

        while (l <= h) 
        {
            int mid = l + (h - l) / 2;
            if (arr[mid] == t)
                return mid;

            if (arr[l] <= arr[mid]) 
            {
                if (t >= arr[l] && t < arr[mid]) 
                {
                    h = mid - 1;  
                }
                else 
                {
                    l = mid + 1;   
                }
            }
            else 
            {
                if (t > arr[mid] && t <= arr[h]) 
                {
                    l = mid + 1;   
                } 
                else 
                {
                    h = mid - 1;  
                }
            }
        }
        return -1;
    }

    public static void main(String args[]) 
    {
        Scanner sc = new Scanner(System.in);
        System.out.print("Enter size of array: ");
        int n = sc.nextInt();
        int arr[] = new int[n];
        int i;

        for (i = 0; i < n; i++)
        {
            System.out.println("Enter an element: ");
            arr[i] = sc.nextInt();
        }

        System.out.print("Enter target value: ");
        int tar = sc.nextInt();

        int result = search(arr, n, tar);
        if (result != -1) 
        {
            System.out.println("target found at index: " + result);
        } 
        else 
        {
            System.out.println("\n" + result);
        }
    }
}

File: DSA/Raima_Rekhi_590013023/Day19Question1.c
#include <stdio.h>

int search_rotated(int arr[],int n,int target);

int main(){
int n,element;

printf("Enter the size of the array:\n");
scanf("%d",&n);
int arr[n];

// Prompt the user to enter the elements of the array
printf("Enter the elements of the array:\n");
for(int i=0;i<n;++i){
scanf("%d",&arr[i]);
}

// Prompt the user to enter the element to be searched in the array
printf("Enter the target element:\n");
scanf("%d",&element);

int index = search_rotated(arr,n,element);

if(index!=-1)
printf("The target element %d found at index: %d\n",element,index);
else
printf("%d\nNo such element found.\n",index);
return 0;
}

// Function to perform binary search on a rotated sorted array
int search_rotated(int arr[],int n,int target){
int left=0,right=n-1;
while(left<=right){
int mid=(right+left)/2;

if(arr[mid]==target) return mid;

if(arr[left]<=arr[mid]){
if(arr[left]<=target && target<arr[mid])
right=mid-1;
else
left=mid+1;
}

else{
if(arr[mid]<target && target<=arr[right])
left=mid+1;
else
right=mid-1;
}
}

return -1; // Target not found
}
